subprojects/core/src/main/java/org/gradle/execution/workgraph/WorkGraphBuilder.java

```diff
@@ -0,0 +1,233 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.execution.workgraph;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+import org.gradle.api.Task;
+import org.gradle.api.internal.TaskInternal;
+import org.gradle.api.internal.project.ProjectInternal;
+import org.gradle.api.internal.tasks.CachingTaskDependencyResolveContext;
+import org.gradle.api.specs.Spec;
+import org.gradle.api.specs.Specs;
+import org.gradle.internal.scheduler.CycleReporter;
+import org.gradle.internal.scheduler.Edge;
+import org.gradle.internal.scheduler.EdgeType;
+import org.gradle.internal.scheduler.Graph;
+import org.gradle.internal.scheduler.Node;
+
+import javax.annotation.Nullable;
+import java.util.ArrayDeque;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static org.gradle.internal.scheduler.EdgeType.AVOID_STARTING_BEFORE_FINALIZED;
+import static org.gradle.internal.scheduler.EdgeType.DEPENDENCY_OF;
+import static org.gradle.internal.scheduler.EdgeType.FINALIZED_BY;
+import static org.gradle.internal.scheduler.EdgeType.MUST_COMPLETE_BEFORE;
+import static org.gradle.internal.scheduler.EdgeType.SHOULD_COMPLETE_BEFORE;
+import static org.gradle.internal.scheduler.NodeState.SHOULD_RUN;
+
+public class WorkGraphBuilder {
+    private static final Graph.LiveEdgeDetector LIVE_EDGE_DETECTOR = new Graph.LiveEdgeDetector() {
+        @Override
+        public boolean isIncomingEdgeLive(Edge edge) {
+            return edge.getType() == DEPENDENCY_OF;
+        }
+
+        @Override
+        public boolean isOutgoingEdgeLive(Edge edge) {
+            return edge.getType() == FINALIZED_BY;
+        }
+    };
+
+    private final Set<Task> requestedTasks = Sets.newLinkedHashSet();
+    private Spec<? super Task> filter = Specs.satisfyAll();
+
+    public void setFilter(@Nullable Spec<? super Task> filter) {
+        if (filter == null) {
+            this.filter = Specs.<Task>satisfyAll();
+        } else {
+            this.filter = filter;
+        }
+    }
+
+    public void addTasks(Iterable<? extends Task> tasks) {
+        Iterables.addAll(requestedTasks, tasks);
+    }
+
+    public WorkGraph build(CycleReporter cycleReporter) {
+        Graph graph = new Graph();
+        Map<Task, TaskNode> taskNodes = Maps.newLinkedHashMap();
+        List<Task> sortedTasks = Lists.newArrayList(requestedTasks);
+        Collections.sort(sortedTasks);
```

Comment:
Why are all the requested tasks sorted here? When requesting `a b` then `a` should be executed first. When requesting `b a` then `b` should be executed first. That's also what happens currently.
I think `addTasks would need to add the tasks in sorted way. For example, when you request `compileJava` in a multiproject build, then the `String` is resolved to all `compileJava` tasks in all the subprojects. Those then need to be sorted and added as one `addTasks` call.